src/Catalyst.Core.Modules.Keystore/KeyChain.cs

```diff
@@ -0,0 +1,497 @@
+using Org.BouncyCastle.Crypto.Digests;
+using Org.BouncyCastle.Crypto.Generators;
+using Org.BouncyCastle.Crypto.Parameters;
+using Org.BouncyCastle.Math;
+using Org.BouncyCastle.OpenSsl;
+using Org.BouncyCastle.Security;
+using Org.BouncyCastle.X509;
+using IKey = Catalyst.Abstractions.Cryptography.IKey;
+
+namespace Catalyst.Core.Modules.Keystore
+{
+    /// <summary>
+    ///   A secure key chain.
+    /// </summary>
+    public partial class KeyChain : IKeyApi
+    {
+        static ILog log = LogManager.GetLogger(typeof(KeyChain));
+
+        private char[] _dek;
+        FileStore<string, EncryptedKey> store;
+
+        /// <summary>
+        ///   Create a new instance of the <see cref="KeyChain"/> class.
+        /// </summary>
+        /// <param name="ipfs">
+        ///   The IPFS Engine associated with the key chain.
+        /// </param>
+        public KeyChain() { }
+
+        FileStore<string, EncryptedKey> Store
+        {
+            get
+            {
+                if (store == null)
+                {
+                    var options = new RepositoryOptions();
+                    var folder = Path.Combine(options.Folder, "keys");
+                    if (!Directory.Exists(folder))
+                        Directory.CreateDirectory(folder);
+                    store = new FileStore<string, EncryptedKey>
+                    {
+                        Folder = folder,
+                        NameToKey = (name) => Encoding.UTF8.GetBytes((string) name).ToBase32(),
+                        KeyToName = (key) => Encoding.UTF8.GetString((byte[]) Base32.Decode(key))
+                    };
+                }
+
+                return store;
+            }
+        }
+
+        /// <summary>
+        ///   The configuration options.
+        /// </summary>
+        public KeyChainOptions Options { get; set; } = new KeyChainOptions();
+
+        /// <summary>
+        ///   Sets the passphrase for the key chain.
+        /// </summary>
+        /// <param name="passphrase"></param>
+        /// <param name="cancel">
+        ///   Is used to stop the task.  When cancelled, the <see cref="TaskCanceledException"/> is raised.
+        /// </param>
+        /// <returns>
+        ///   A task that represents the asynchronous operation.
+        /// </returns>
+        /// <exception cref="UnauthorizedAccessException">
+        ///   When the <paramref name="passphrase"/> is wrong.
+        /// </exception>
+        /// <remarks>
+        ///   The <paramref name="passphrase"/> is used to generate a DEK (derived encryption
+        ///   key).  The DEK is then used to encrypt the stored keys.
+        ///   <para>
+        ///   Neither the <paramref name="passphrase"/> nor the DEK are stored.
+        ///   </para>
+        /// </remarks>
+        public async Task SetPassphraseAsync(SecureString passphrase,
+            CancellationToken cancel = default(CancellationToken))
+        {
+            // TODO: Verify DEK options.
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Complete the task associated to this 'TODO' comment.](https://app.codacy.com/app/catalyst-network/Catalyst.Framework/pullRequest?prid=4612924)